HTTPCLIENT-1625 Completely overhaul GSS-API-based authentication backend

stoty opened the pull request.

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java:
```diff
@@ -60,69 +60,167 @@
-@Deprecated
-public abstract class GGSSchemeBase implements AuthScheme {
+// FIXME The class name looks like a Typo. Rename in 6.0 ?
+public abstract class GGSSchemeBase implements AuthSchemeV2 {
```
It is a typo, I reported this. We can safely rename it.

stoty replied:
Done

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java:
```diff
@@ -60,69 +60,167 @@
+
+    // The GSS spec does not specify how long the conversation can be. This should be plenty.
+    // Realistically, we get one initial token, then one maybe one more for mutual authentication.
+    private static final int MAX_GSS_CHALLENGES = 3;
```
Why? I don't see a need and have never seen such a boundary. The context is complete when it is complete. In the past 15 years I haven't seen more than one roundtrip for SPNEGO. This can only happen when IAKerb is in place...

stoty replied:
There is an explicit test to avoid the infinite conversation in the old code, and logic to only ever allow one challenge.

The one challenge case is obviously not working for mutual auth, which relies on processing a second token.

I could argue that this check also protects from malicious DOS attacks.

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java:
```diff
@@ -60,69 +60,167 @@
+                throw new AuthenticationException(gsse.getMessage(), gsse);
+            }
+            // other error
+            throw new AuthenticationException(gsse.getMessage(), gsse);
```
Won't those duplicate the message?

stoty replied:
Not really.
But the Auth. execption would have the same message as the cause.
We can write some new message for the Auth. exception, I just really hate when a library swallows the original exception instead of wrapping it.

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java:
```diff
@@ -138,7 +236,9 @@ protected byte[] generateGSSToken(
         final GSSManager manager = getManager();
         final GSSName serverName = manager.createName(serviceName + "@" + authServer, GSSName.NT_HOSTBASED_SERVICE);
```
I must admit that `authServer` is a horrible variable name.

stoty replied:
fixed

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/GGSSchemeBase.java:
```diff
@@ -138,7 +236,9 @@ protected byte[] generateGSSToken(
         final GSSManager manager = getManager();
         final GSSName serverName = manager.createName(serviceName + "@" + authServer, GSSName.NT_HOSTBASED_SERVICE);
```
Well, that is not a server name, but a SPN...

stoty replied:
This is unchanged, but I will rename it.

stoty replied:
What would you prefer ? 'peer' ? 'principal' ? 'spn' ?

michael-o replied:
peerName sounds reasonable and it also reflects the same term in the TLS code, i.e., peer certificate.

stoty replied:
I went with spn for now.

stoty replied:
renamed to peerName

michael-o replied:
> I went with spn for now.

SPN is a Kerberos term, not GSS-API. `peerName` is just acceptable.

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/KerberosScheme.java:
```diff
@@ -40,14 +40,7 @@
- * @see BasicScheme
- * @see BearerScheme
  */
-@Deprecated
```
Will go

michael-o commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/SPNegoScheme.java:
```diff
@@ -41,14 +41,7 @@
  */
-@Deprecated
 @Experimental
 public class SPNegoScheme extends GGSSchemeBase {
```
Should also go in favor of `SpnegoScheme`

stoty replied:
OK

michael-o reviewed (changes requested):
First, incomplete review.

stoty commented:
Thanks for the initial review @michael-o .

Regarding the renames and changing the defaults, I have specifically tried to touch as little as possible because much of the GSS/Kerberos code was not removed from codebase when disabling SPNEGO, which I assumed was for brackwards compatibility reasons.

I'm fine with your proposed changes, but those may theoretically cause issues with pre-5.3 apps, and will cause the api compare plugin to go wild.

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/auth/AuthSchemeV2.java:
```diff
@@ -0,0 +1,102 @@
+ *
+ * @since 5.5
+ */
+public interface AuthSchemeV2 extends AuthScheme {
```
OK

stoty replied:
done

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/auth/KerberosConfig.java:
```diff
@@ -53,25 +49,28 @@ public enum Option {
-    private final Option stripPort;
-    private final Option useCanonicalHostname;
-    private final Option requestDelegCreds;
+    private final Option stripPort; //Effective default is ENABLE
```
That's certainly possible, but that would introduce an incompatibility with the 5.2 behaviour.

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/auth/KerberosConfig.java:
```diff
@@ -53,25 +49,28 @@ public enum Option {
+    private final Option stripPort; //Effective default is ENABLE
+    private final Option useCanonicalHostname; //Effective default is ENABLE
+    private final Option requestDelegCreds; //Effective default is DISABLE
+    private final Option requestMutualAuth; //Effective default is DISABLE
```
All I can came up with is:
- GSS does not default to mutual auth
- Backwards compatibility

I would at least leave an option to disable this, so that there is an out for very broken cases.
Of course there is also a case for forcing users to clean up their potentially broken setup.

stoty replied:
I have looked at this, and the current API is really built around using the gss library default settings.

Since we are making incompatible changes anyway, if we want to set non-default gss options by default, (and keep them configurable), then we should remove the default options, and use simple booleans with explicit true/false defaults.

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/auth/StandardAuthScheme.java:
```diff
@@ -66,20 +66,12 @@ private StandardAuthScheme() {
-     * supported. Consider using Basic or Bearer authentication with TLS instead.
      */
-    @Deprecated
     public static final String KERBEROS = "Kerberos";
```
OK

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/DefaultAuthenticationStrategy.java:
```diff
@@ -68,6 +68,8 @@ public class DefaultAuthenticationStrategy implements AuthenticationStrategy {
 
     private static final List<String> DEFAULT_SCHEME_PRIORITY =
         Collections.unmodifiableList(Arrays.asList(
+                StandardAuthScheme.SPNEGO,
```
OK, having to define a new strategy for clients is no big deal.

Should we have one in the code that can be used, or should we leave it to user to subclass ?

stoty replied:
Done.

michael-o commented:
> Thanks for the initial review @michael-o .
> 
> Regarding the renames and changing the defaults, I have specifically tried to touch as little as possible because much of the GSS/Kerberos code was not removed from codebase when disabling SPNEGO, which I assumed was for brackwards compatibility reasons.
> 
> I'm fine with your proposed changes, but those may theoretically cause issues with pre-5.3 apps, and will cause the api compare plugin to go wild.

Will engage next week again.

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/auth/AuthScheme.java:
```diff
@@ -86,6 +86,10 @@
  * and in FAILED state.
  * </p>
+ * <p>
+ * This interface cannot correctly handle some authentication methods, like SPENGO.
```
fixed.

stoty commented:
Can you take another look at the current patch @michael-o ?
I have implemented most of your suggestions, but I had a few more questions comments regarding KerberosConfig.

michael-o commented:
> Can you take another look at the current patch @michael-o ? I have implemented most of your suggestions, but I had a few more questions comments regarding KerberosConfig.

Will pick end of this week or next week.

stoty commented:
Rebased to current master

michael-o commented:
@ok2c @stoty I gave this another conceptual thought: We said that we are going to remove the current code for good, but this PR reuses it and makes it partially undeprecated. For me, this is not straight forward. My gut feeling says that we should remove completely *first*  and the @stoty can readd the code he thinks is required to satisfy this PR. WDYT?

ok2c commented:
@michael-o We cannot remove old implementation unless we are willing to go to the major release cycle and release these changes as 6.0. I suggest that the old classes are left deprecated as there might still be users that rely on them and new implementation is added in parallel (with different names or in a different package).

michael-o commented:
> @michael-o We cannot remove old implementation unless we are willing to go to the major release cycle and release these changes as 6.0. I suggest that the old classes are left deprecated as there might still be users that rely on them and new implementation is added in parallel (with different names or in a different package).

I see, than we need new packages names. Agreed.

stoty commented:
I have reverted the removals.

I don't see what alternative package name we could use instead of "org.apache.hc.client5.http.auth"
Using another package might also break package level visibility.

Maybe I could prefix the new classes with "Mutual"  ?
i.e 
MutualGssSchemeBase
MutualSpengoScheme 

The other classes are strictly backwards compatible, like KerberosCredentials, and KerberosConfig.
Should I duplicate those as well ?

michael-o commented:
> I have reverted the removals.
> 
> I don't see what alternative package name we could use instead of "org.apache.hc.client5.http.auth" Using another package might also break package level visibility.
> 
> Maybe I could prefix the new classes with "Mutual" ? i.e MutualGssSchemeBase MutualSpengoScheme
> 
> The other classes are strictly backwards compatible, like KerberosCredentials, and KerberosConfig. Should I duplicate those as well ?

I'll get back to this question this week, have some serious trouble at $work I need to solve first.

stoty commented:
I don't think that removing deprecation is problem, especially as these classes were deprecated without providing a replacement, not because they were replaced by something.

ok2c commented:
> I don't think that removing deprecation is problem, especially as these classes were deprecated without providing a replacement, not because they were replaced by something.

@stoty Un-deprecate those classes that can be kept fully backward compatible and create new classes for those that cannot

stoty commented:
I have made the discussed changes.

michael-o commented:
> @stoty @michael-o Folks, what is the reason this PR is stuck? @stoty Please rebase the latest off master to pick up Docker based compatibility tests. Feel free to disable `japicmp` but all integration must pass with your changes.

I am busy, can only peek once in a while. Hopefully the upcoming weeks will be better.

ok2c commented:
@stoty If you want to expedite the review process we could do the following. Once you are more or less happy with the state of things, submit changes to `HttpAutheticator` and `AuthScheme` in a separate PR and will happily review them. Kerberos specific code will have to wait until @michael-o has time to review it.

stoty commented:
Thank you @ok2c.

I have rebased the patch and temporarily disabled japcmp.
I have also added and @Internal annotation to HttpAuthenticator, I cannot imagine realistic user code calling that directly.
(though it does not seem to apply retroactively)

I will try to make the time to split the PR.

stoty commented:
I don't think that duplicating HttpAuthenticator is a good idea.

For the authentication shemes, a case could be made for duplicating them, as those are pluggable components.

However, HttpAuthenticator is a purely internal class, and no sane external code is expected to call it.

The only way I could think that HttpAuthenticator would be called by external code, is if someone fully re-implemented a new (Closable...)Client class (and the corresponding ...Exec class ), which is not something that I think should be covered by backwards compatibility.

The end result would be replacing HttpAuthenticator with HttpAuthenticator2, and having a dangling HttpAuthenticator, which doesn't work with the other classes anyway (and it would possibly have to be modified anyway just to compile).

ok2c commented:
@stoty Please disable `japicmp` for now and make changes to `HttpAuthenticator ` you deem necessary. I will take it from there.

stoty commented:
Thank you.
I have already done that.

stoty commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/auth/HttpAuthenticator.java:
```diff
@@ -186,52 +202,109 @@ public boolean updateAuthState(
                             authExchange.setState(AuthExchange.State.FAILURE);
-                            return false;
+                            if (isChallengeExpected) {
+                                throw ex;
```
Mutual auth case, throw exception, because the server has sent a 200 or equvalent code, and we will simply return the response to the client otherwise, even if the AuthExchange.STATE is Failure.

stoty reviewed (commented):
I have add some comments to the most important changes

stoty commented:
@michael-o, @ok2c 

Now that the festive sesason is over, are you able to find the time for reviews ?

On my side, I still need to implement the requested new config interface for the new auth methods, but that should not affect the rest of the changes.

ok2c commented:
> @michael-o, @ok2c
> 
> Now that the festive sesason is over, are you able to find the time for reviews ?
> 
> On my side, I still need to implement the requested new config interface for the new auth methods, but that should not affect the rest of the changes.

@stoty What shall I say? I will repeat the same thing I have said you before. If you split this change-set into two: one with auth API changes only and another one with Kerberos auth, I could help you get the former incorporated into HttpClient. Otherwise you are fully at @michael-o's mercy.

stoty commented:
So the first patch would be most of this patch, without the actual MutualSPnego implementation, and corresponding new test, while the second one would add the new MutualSpnego implementation class, and the new test class (plus some new comments that refer to them) ?

ok2c commented:
> So the first patch would be most of this patch, without the actual MutualSPnego implementation, and corresponding new test, while the second one would add the new MutualSpnego implementation class hierarchy, and the new test class (plus some new comments that refer to them) ?

@stoty Correct. This way, in the worst case scenario one would be able to plug-in a custom MutualSpnego implementation even if the auth scheme is not supported by the project.

stoty commented:
Opened #612 for the first part.

michael-o commented:
@stoty Can you rebase please?

stoty commented:
#615 is the rebased PR top of current master @michael-o .

I am closing this now.
